lib/puppet/pops/loader/task_instantiator.rb

```diff
@@ -3,66 +3,95 @@
+        metadata['implementations'].map do |impl|
+          path = executables.find { |real_impl| File.basename(real_impl) == impl['name'] }
+          if path
+            { "name" => impl['name'], "requirements" => impl.fetch('requirements', []), "path" => path }
+          else
+            raise ArgumentError, _("Task metadata for task %{name} specifies missing implementation %{implementation}") %
+              { name: name, implementation: impl['name'] }
+          end
+        end
+      else
+        # If 'implementations' is the wrong type, we just pass it through and
+        # let the task type definition reject it.
+        metadata['implementations']
+      end
+    # If implementations isn't defined, then we use executables matching the
+    # task name, and only one may exist.
+    else
+      implementations = executables.select { |impl| File.basename(impl, '.*') == basename }
+      if implementations.empty?
+        raise ArgumentError, _('No source besides task metadata was found in directory %{directory} for task %{name}') %
+          { name: name, directory: directory }
+      elsif implementations.length > 1
+        # XXX This message sucks
+        raise ArgumentError, _("Multiple executables were found in directory %{directory} for task %{name}, without differentiating metadata") %
+          { name: name, directory: implementations[0] }
+      end
+
+      [{ "name" => File.basename(implementations.first), "path" => implementations.first, "requirements" => [] }]
+    end
+  end
+
+  def self.create(loader, typed_name, source_refs)
+    name = typed_name.name
+    basename = typed_name.name_parts[1] || 'init'
+    dirname = File.dirname(source_refs[0])
+    metadata_files, executables = source_refs.partition { |source_ref| source_ref.end_with?('.json') }
+    metadata_file = metadata_files.find { |source_ref| File.basename(source_ref, '.json') == basename }
+
+    metadata = load_metadata(loader, metadata_file)
+
+    implementation_metadata = validate_implementations(typed_name, dirname, metadata, executables)
+
     arguments = {
       'name' => name,
-      'executable' => task_source
+      'implementations' => implementation_metadata
     }
-    hash.each_pair do |key, value|
-      if 'parameters' == key || 'output' == key
-        ps = {}
-        value.each_pair do |k, v|
-          pd = v.dup
-          t = v['type']
-          pd['type'] = t.nil? ? Types::TypeFactory.data : Types::TypeParser.singleton.parse(t)
-          ps[k] = pd
+
+    begin
+      metadata.each_pair do |key, value|
+        if %w[parameters output].include?(key)
+          ps = {}
+          value.each_pair do |k, v|
+            pd = v.dup
+            t = v['type']
+            pd['type'] = t.nil? ? Types::TypeFactory.data : Types::TypeParser.singleton.parse(t)
+            ps[k] = pd
+          end
+          value = ps
         end
-        value = ps
+        arguments[key] = value unless arguments.key?(key)
```

Comment:
Why do we need to check? Is this fixing a different bug?

Reply:
We compute a few keys on our own (`name` and `implementations`), and this prevents those 
values from being overridden with the value from the metadata.